### PR TITLE
Merge of initial 7-BM customizations

### PR DESCRIPTION
**Pull request by aps-7bm**
I have done some initial testing of my customizations for 7-BM.  One of the main differences from the 2-BM version is that I am calculating the PSO parameters in the tomoscan_7bm.py file, rather than using the PSOFly infrastructure set up by Tim Mooney from BCDA.  The logic is based on Tim's PSOFly, and mostly sends commands directly to the Aerotech Ensemble driver through the asyn record set up for PSOFly.

### CODE REVIEW

**Comment by decarlof** [comment]
@aps-7bm I would remove the print() in the tomoscan.py and perhaps undo the changes in configure/RELEASE as this is happening when following the [Install directions](https://tomoscan.readthedocs.io/en/latest/install.html).
I like your PSO handling better, we should do the same at 2-BM and 32-ID and perhaps consider merging at one point 2-BM and 7-BM. 

@MarkRivers @aps-7bm please hold merging as I am adding a doc update to include the 7-BM additions (will be done later today)

**Comment by aps-7bm** [comment]
I added the screenshot of the MEDM screen, as requested.  I've kept working on this and have added a couple more features.  I added a different exposure time to the base class and screen, since that seemed like something that would be more generally useful.  I reworked my MEDM files quite a bit, since the 7BM screen was getting monstrous.  I also changed the logic a bit so that end_scan would get called in a finally: statement, so the end of scan tasks get done even if there is an exception thrown during the scan.

**Comment by xray-imaging** [comment]
@aps-7bm all looks good to me. @MarkRivers may you please double check tomoscan/tomoscan.py changes?

File: tomoScanApp/Db/tomoScan.template [hunk 1]
```diff
@@ -199,6 +199,17 @@ record(ao, "$(P)$(R)ExposureTime")
    field(PREC,  "3")
 }
 
+record(ao, "$(P)$(R)BrightExposureTime")
```
Comment (MarkRivers) [inline tomoScanApp/Db/tomoScan.template]: Everywhere else the term "FlatField" is used, not Briight.  This should be changed to FlatFieldExposureTime or if that is too long "FlatExposureTime".  Similarly it should be "DifferentFlatExposureTime".

File: tomoScanApp/Db/tomoScan_settings.req [hunk 2]
```diff
@@ -68,6 +68,8 @@ $(P)$(R)SampleOutY
 # Exposure time
 #################
 $(P)$(R)ExposureTime
+$(P)$(R)BrightExposureTime
```
Comment (MarkRivers) [inline tomoScanApp/Db/tomoScan_settings.req]: Change Bright to Flat here.

File: tomoscan/tomoscan.py [hunk 3]
```diff
@@ -489,7 +493,27 @@ def set_exposure_time(self, exposure_time=None):
 
         if exposure_time is None:
             exposure_time = self.epics_pvs['ExposureTime'].value
-        self.epics_pvs['CamAcquireTime'].put(exposure_time)
+        self.epics_pvs['CamAcquireTime'].put(exposure_time, wait=True, timeout = 10.0)
```
Comment (MarkRivers) [inline tomoscan/tomoscan.py:496]: Chnage Bright to Flat

File: tomoScanApp/op/adl/tomoScan.adl [hunk 4]
```diff
@@ -1,14 +1,14 @@
 
 file {
```
Comment (MarkRivers) [inline tomoScanApp/op/adl/tomoScan.adl:2]: I downloaded and viewed the new version of tomoScan.adl in medm.  It has a number of problems:
- The total height of the display is wrong, the bottom is cut off.  It was previously 825 but was changed to 803 for some reason.
- All widget X and Y positions should be a multiple of 5.  That keeps them correctly aligned. 
- Text labels should be positioned immediately adjacent to the widget they describe using "Space Evenly/Horizontal"
- Text labels should be exactly the size required by the text using "Size/Text to Contents".

File: tomoScanApp/Db/tomoScan.template (hunk 1 above)
Comment (MarkRivers) [reply tomoScanApp/Db/tomoScan.template]: Actually I think it should be "DifferentFlatExposure" without Time.  The PV name is getting near the maximum length depending on the prefixes.

**Comment by MarkRivers** [comment]
I have made my suggested edits with medm to the adl file.  It is attached here with .txt on the end.
[tomoScan.adl.txt](https://github.com/tomography/tomoscan/files/5268222/tomoScan.adl.txt)

**Comment by aps-7bm** [comment]
I made those changes and committed them, those were good points.  I should
have looked more carefully at your naming.  The MEDM screen looks better; I
frankly stink at making screens that look good, so I appreciate the
advice.  I have one question.  In the flat field part of the screen, all of
the labels on the far left are spaced a few pixels away from the fields to
their right except the Flat Exposure one.  Should we align it with the
labels above it, either by moving it left or the other labels right?

**Comment by MarkRivers** [comment]
>  In the flat field part of the screen, all of the labels on the far left are spaced a few pixels away from the fields to their right except the Flat Exposure one. 

I don't see that, I just downloaded your latest version.  "Flat exposure" has X=35, and is 13 characters.  Each character takes 10 pixels in that font, so it ends at pixel 165.  The Menu widget to its right starts at pixel 170, so there are 5 pixels of space between them, which is what "Space evenly" does. 

Perhaps you are using a different font alias file from me?  This is what the screen looks like for me:

image

There was one minor problem, those 2 widgets were 1 pixel too low.  That can be fixed with this patch:
```
corvette:tomoScanApp/op/adl>git diff .
diff --git a/tomoScanApp/op/adl/tomoScan.adl b/tomoScanApp/op/adl/tomoScan.adl
index 164a3c6..31f5eaf 100644
--- a/tomoScanApp/op/adl/tomoScan.adl
+++ b/tomoScanApp/op/adl/tomoScan.adl
@@ -1238,7 +1238,7 @@ menu {
 text {
        object {
                x=35
-               y=336
+               y=335
                width=130
                height=20
        }
@@ -1251,7 +1251,7 @@ text {
 menu {
        object {
                x=170
-               y=336
+               y=335
                width=100
                height=20
        }
```

**Comment by aps-7bm** [comment]
Mark,
    This is what it looks like for me on my beamline workstation.

[image: image.png]

**Comment by MarkRivers** [comment]
@aps-7bm your screen-shot did not get uploaded properly.  You can drag a png file into the window, or copy and paste.

**Comment by aps-7bm** [comment]
OK, that's weird.  When I look at it now, it looks just like yours.  Maybe
something didn't update properly earlier.  I'm good now, I think.  Any
other suggestions?

**Comment by MarkRivers** [comment]
It's OK to merge now in my opinion.

**Comment by aps-7bm** [comment]
I just pushed that shift of the two components in the MEDM screen.

**Comment by decarlof** [comment]
ok

**Pull request merged**